start does not work with yarn workspaces

### Preflight Checklist
<!-- Please ensure you've completed the following steps by replacing [ ] with [x]-->

* [X] I have read the [contribution documentation](https://github.com/electron-userland/electron-forge/blob/master/CONTRIBUTING.md) for this project.
* [X] I agree to follow the [code of conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project follows, as appropriate.
* [X] I have searched the issue tracker for a bug that matches the one I want to file, without success.

### Issue Details

* **Electron Forge Version:**
  * 6.0.0-beta.47
* **Electron Version:**
  * v7.1.7
* **Operating System:**
  * macOS 10.14.6 (18G95)
* **Last Known Working Electron Forge version:**:
  * none

### Expected Behavior
No error

### Actual Behavior

![CleanShot 2020-01-01 at 02 31 01](https://user-images.githubusercontent.com/14089557/71632022-c6048180-2c3e-11ea-844b-cd1e1a75b583.png)

### To Reproduce
<!--
Your best chance of getting this bug looked at quickly is to provide a MINIMAL REPOSITORY that can be cloned and run. Also include:
* if you are using the `electron-packager` CLI: the command line arguments you are passing
* if you are using the API the parameters are you passing to the `packager()` function
-->

- This is the error repository: https://github.com/kamontat/KingKnight/tree/issues/yarn-workspace
- In order to run the project:
    - run `yarn install:all`
    - run `yarn compile`
    - run `yarn start`

### Additional Information

Error log:

```
@kk/app: $ electron-forge start
@kk/app: WARNING: DEBUG environment variable detected.  Progress indicators will be sent over electron-forge:lifecycle
@kk/app: 2019-12-31T19:33:52.010Z electron-forge:async-ora Process Started: Checking your system
@kk/app: 2019-12-31T19:33:52.011Z electron-forge:check-system checking system, create ~/.skip-forge-system-check to stop doing this
@kk/app: 2019-12-31T19:33:52.269Z electron-forge:async-ora Process Succeeded: Checking your system -- after 259ms
@kk/app: WARNING: DEBUG environment variable detected.  Progress indicators will be sent over electron-forge:lifecycle
@kk/app: 2019-12-31T19:33:52.601Z electron-forge:async-ora Process Started: Locating Application
@kk/app: 2019-12-31T19:33:52.602Z electron-forge:project-resolver searching for project in: /Users/kchantrachir/Desktop/Personal/KingKnight/packages/app
@kk/app: 2019-12-31T19:33:52.603Z electron-forge:project-resolver electron-forge compatible package.json found in /Users/kchantrachir/Desktop/Personal/KingKnight/packages/app/package.json
@kk/app: 2019-12-31T19:33:52.603Z electron-forge:async-ora Process Succeeded: Locating Application -- after 2ms
@kk/app: 2019-12-31T19:33:52.606Z electron-forge:require-search searching [
@kk/app:   '@electron-forge/plugin-webpack',
@kk/app:   '/Users/kchantrachir/Desktop/Personal/KingKnight/packages/app/@electron-forge/plugin-webpack',
@kk/app:   '/Users/kchantrachir/Desktop/Personal/KingKnight/packages/app/node_modules/@electron-forge/plugin-webpack'
@kk/app: ] relative to /Users/kchantrachir/Desktop/Personal/KingKnight/packages/app
@kk/app: 2019-12-31T19:33:52.607Z electron-forge:require-search testing @electron-forge/plugin-webpack
@kk/app: 2019-12-31T19:33:52.848Z electron-forge:plugin:webpack hooking process events
@kk/app: 2019-12-31T19:33:52.850Z electron-forge:async-ora Process Started: Preparing native dependencies
@kk/app: 2019-12-31T19:33:52.856Z electron-forge:async-ora Process Succeeded: Preparing native dependencies -- after 6ms
@kk/app: 2019-12-31T19:33:52.856Z electron-forge:plugins plugin: "webpack" has taken control of the start command
@kk/app: 2019-12-31T19:33:52.861Z electron-forge:async-ora Process Started: Compiling Main Process Code
@kk/app: Starting type checking service...
@kk/app: Using 1 worker with 2048MB memory limit
@kk/app: 2019-12-31T19:33:56.618Z electron-forge:async-ora Process Succeeded: Compiling Main Process Code -- after 3.8s
@kk/app: 2019-12-31T19:33:56.618Z electron-forge:async-ora Process Started: Launch Dev Servers
@kk/app: 2019-12-31T19:33:56.635Z electron-forge:async-ora Process Succeeded: Launch Dev Servers -- after 17ms
@kk/app: 2019-12-31T19:33:56.635Z electron-forge:async-ora Process Started: Compiling Preload Scripts
@kk/app: 2019-12-31T19:33:56.635Z electron-forge:async-ora Process Succeeded: Compiling Preload Scripts -- after 0ms
@kk/app:
@kk/app: An unhandled rejection has occurred inside Forge:
@kk/app: Error: Cannot find module '/Users/kchantrachir/Desktop/Personal/KingKnight/packages/app/node_modules/electron'
@kk/app: Require stack:
@kk/app: - /Users/kchantrachir/Desktop/Personal/KingKnight/node_modules/@electron-forge/core/dist/api/start.js
@kk/app: - /Users/kchantrachir/Desktop/Personal/KingKnight/node_modules/@electron-forge/core/dist/api/index.js
@kk/app: - /Users/kchantrachir/Desktop/Personal/KingKnight/node_modules/@electron-forge/cli/dist/electron-forge-start.js
@kk/app:     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:797:15)
@kk/app:     at Function.Module._load (internal/modules/cjs/loader.js:690:27)
@kk/app:     at Module.require (internal/modules/cjs/loader.js:852:19)
@kk/app:     at require (internal/modules/cjs/helpers.js:74:18)
@kk/app:     at forgeSpawn (/Users/kchantrachir/Desktop/Personal/KingKnight/node_modules/@electron-forge/core/src/api/start.ts:81:26)
@kk/app:     at processTicksAndRejections (internal/process/task_queues.js:93:5)
@kk/app:     at forgeSpawnWrapper (/Users/kchantrachir/Desktop/Personal/KingKnight/node_modules/@electron-forge/core/src/api/start.ts:121:21)
@kk/app:     at /Users/kchantrachir/Desktop/Personal/KingKnight/node_modules/@electron-forge/cli/src/electron-forge-start.ts:59:19 {
@kk/app:   code: 'MODULE_NOT_FOUND',
@kk/app:   requireStack: [
@kk/app:     '/Users/kchantrachir/Desktop/Personal/KingKnight/node_modules/@electron-forge/core/dist/api/start.js',
@kk/app:     '/Users/kchantrachir/Desktop/Personal/KingKnight/node_modules/@electron-forge/core/dist/api/index.js',
@kk/app:     '/Users/kchantrachir/Desktop/Personal/KingKnight/node_modules/@electron-forge/cli/dist/electron-forge-start.js'
@kk/app:   ]
@kk/app: }
@kk/app:
@kk/app: Electron Forge was terminated. Location:
@kk/app: {}
@kk/app: 2019-12-31T19:33:56.644Z electron-forge:plugin:webpack handling process exit with: { cleanup: true }
@kk/app: 2019-12-31T19:33:56.644Z electron-forge:plugin:webpack cleaning webpack watcher
@kk/app: 2019-12-31T19:33:56.645Z electron-forge:plugin:webpack cleaning http server
@kk/app: 2019-12-31T19:33:56.645Z electron-forge:plugin:webpack stopping logger
```

This is a code that might relate to this issue

https://github.com/electron-userland/electron-forge/blob/9b8a2a6acbccf280472b0a378451d6520322a542/packages/api/core/src/api/start.ts#L81

<!--
Add any other context about the problem here.

For example:
* Console output when you run your `electron-forge` command with the environment variable
  `DEBUG=electron-forge:*`. (Instructions on how to do so
  [here](https://www.npmjs.com/package/debug#usage)). Please include the stack trace if
  one exists.
* Command line arguments you are passing to `electron-forge` (e.g.,
  `electron-forge make --not-a-real-flag`)
* The `config.forge` data in `package.json` or `forge.config.js` in use
* A failing minimal testcase (with a link to the code) or detailed steps to reproduce the problem.
  Using `electron-forge init` is a good starting point, if that is not the source of your problem.

--- a/packages/api/core/src/api/start.ts
+++ b/packages/api/core/src/api/start.ts
@@ -11,7 +11,7 @@ import rebuild from '../util/rebuild';
 import resolveDir from '../util/resolve-dir';
 import getForgeConfig from '../util/forge-config';
 import { runHook } from '../util/hook';
-import { getElectronVersion } from '../util/electron-version';
+import { getElectronModulePath, getElectronVersion } from '../util/electron-version';
 
 export { StartOptions };
 
@@ -78,7 +78,7 @@ export default async ({
 
     if (!electronExecPath) {
       // eslint-disable-next-line import/no-dynamic-require, global-require
-      electronExecPath = require(path.resolve(dir, 'node_modules/electron'));
+      electronExecPath = require((await getElectronModulePath(dir, packageJSON)) || path.resolve(dir, 'node_modules/electron'));
     }
 
     const spawnOpts = {

--- a/packages/api/core/src/util/electron-version.ts
+++ b/packages/api/core/src/util/electron-version.ts
@@ -46,23 +46,55 @@ export class PackageNotFoundError extends Error {
   }
 }
 
-export async function getElectronVersion(dir: string, packageJSON: any): Promise<string> {
+function getElectronModuleName(packageJSON: any): string {
   if (!packageJSON.devDependencies) {
     throw new Error('package.json for app does not have any devDependencies'.red);
   }
+
   const packageName = electronPackageNames.find((pkg) => packageJSON.devDependencies[pkg]);
   if (packageName === undefined) {
     throw new Error('Could not find any Electron packages in devDependencies');
   }
 
+  return packageName;
+}
+
+async function getElectronPackageJSONPath(
+  dir: string,
+  packageName: string,
+): Promise<string | undefined> {
+  const nodeModulesPath = await determineNodeModulesPath(dir);
+  if (!nodeModulesPath) {
+    throw new PackageNotFoundError(packageName, dir);
+  }
+  const electronPackageJSONPath = path.join(nodeModulesPath, packageName, 'package.json');
+  if (await fs.pathExists(electronPackageJSONPath)) {
+    return electronPackageJSONPath;
+  }
+
+  return undefined;
+}
+
+export async function getElectronModulePath(
+  dir: string,
+  packageJSON: any,
+): Promise<string | undefined> {
+  const moduleName = await getElectronModuleName(packageJSON);
+  const packageJSONPath = await getElectronPackageJSONPath(dir, moduleName);
+  if (packageJSONPath) {
+    return path.dirname(packageJSONPath);
+  }
+
+  return undefined;
+}
+
+export async function getElectronVersion(dir: string, packageJSON: any): Promise<string> {
+  const packageName = getElectronModuleName(packageJSON);
+
   let version = packageJSON.devDependencies[packageName];
   if (!semver.valid(version)) { // It's not an exact version, find it in the actual module
-    const nodeModulesPath = await determineNodeModulesPath(dir);
-    if (!nodeModulesPath) {
-      throw new PackageNotFoundError(packageName, dir);
-    }
-    const electronPackageJSONPath = path.join(nodeModulesPath, packageName, 'package.json');
-    if (await fs.pathExists(electronPackageJSONPath)) {
+    const electronPackageJSONPath = await getElectronPackageJSONPath(dir, packageName);
+    if (electronPackageJSONPath) {
       const electronPackageJSON = await fs.readJson(electronPackageJSONPath);
       // eslint-disable-next-line prefer-destructuring
       version = electronPackageJSON.version;

--- a/packages/api/core/test/fast/electron-version_spec.ts
+++ b/packages/api/core/test/fast/electron-version_spec.ts
@@ -1,6 +1,8 @@
 import { expect } from 'chai';
+import fs from 'fs-extra';
+import os from 'os';
 import path from 'path';
-import { getElectronVersion, updateElectronDependency } from '../../src/util/electron-version';
+import { getElectronModulePath, getElectronVersion, updateElectronDependency } from '../../src/util/electron-version';
 import { devDeps, exactDevDeps } from '../../src/api/init-scripts/init-npm';
 import { hasYarn } from '../../src/util/yarn-or-npm';
 
@@ -84,5 +86,44 @@ describe('getElectronVersion', () => {
     } else {
       expect(getElectronVersion(fixtureDir, packageJSON)).to.eventually.be.rejectedWith('Cannot find the package');
     }
+  });
+});
+
+describe('getElectronModulePath', () => {
+  let tempDir: string;
+  before(async () => {
+    tempDir = await fs.mkdtemp(path.join(os.tmpdir(), 'electron-forge-test-'));
+  });
+
+  it('fails without devDependencies', () => expect(getElectronModulePath('', {})).to.eventually.be.rejectedWith('does not have any devDependencies'));
+
+  it('fails without electron devDependencies', () => expect(getElectronModulePath('', { devDependencies: {} })).to.eventually.be.rejectedWith('Electron packages in devDependencies'));
+
+  it('fails with no electron installed', async () => {
+    const fixtureDir = path.resolve(__dirname, '..', 'fixture', 'dummy_app');
+    await fs.copy(fixtureDir, tempDir);
+    return expect(getElectronModulePath(tempDir, { devDependencies: { electron: '^4.0.2' } })).to.eventually.be.rejectedWith('Cannot find the package');
+  });
+
+  it('works with electron', () => {
+    const fixtureDir = path.resolve(__dirname, '..', 'fixture', 'non-exact');
+    return expect(getElectronModulePath(fixtureDir, { devDependencies: { electron: '^4.0.2' } })).to.eventually.equal(path.join(fixtureDir, 'node_modules', 'electron'));
+  });
+
+  it('works with yarn workspaces', async () => {
+    const workspaceDir = path.resolve(__dirname, '..', 'fixture', 'yarn-workspace');
+    const fixtureDir = path.join(workspaceDir, 'packages', 'subpackage');
+    const packageJSON = {
+      devDependencies: { electron: '^4.0.4' },
+    };
+    if (hasYarn()) {
+      expect(await getElectronModulePath(fixtureDir, packageJSON)).to.be.equal(path.join(workspaceDir, 'node_modules', 'electron'));
+    } else {
+      expect(getElectronModulePath(fixtureDir, packageJSON)).to.eventually.be.rejectedWith('Cannot find the package');
+    }
+  });
+
+  after(async () => {
+    await fs.remove(tempDir);
   });
 });

--- a/packages/api/core/test/fast/start_spec.ts
+++ b/packages/api/core/test/fast/start_spec.ts
@@ -19,15 +19,20 @@ describe('start', () => {
     spawnStub = sinon.stub();
     shouldOverride = false;
     packageJSON = require('../fixture/dummy_app/package.json');
+    const electronPath = path.resolve(__dirname, 'node_modules/electron');
 
     start = proxyquire.noCallThru().load('../../src/api/start', {
+      '../util/electron-version': {
+        getElectronModulePath: () => Promise.resolve(electronPath),
+        getElectronVersion: () => Promise.resolve('1.0.0'),
+      },
       '../util/forge-config': async () => ({
         pluginInterface: {
           overrideStartLogic: async () => shouldOverride,
           triggerHook: async () => false,
         },
       }),
-      [path.resolve(__dirname, 'node_modules/electron')]: 'fake_electron_path',
+      [electronPath]: 'fake_electron_path',
       '../util/resolve-dir': async (dir: string) => resolveStub(dir),
       '../util/read-package-json': {
         readMutatedPackageJson: () => Promise.resolve(packageJSON),

